crates/ibc/src/core/ics02_client/handler/upgrade_client.rs

```diff
@@ -35,22 +35,50 @@ where
 {
     let MsgUpgradeClient { client_id, .. } = msg;
 
-    // Read client state from the host chain store.
+    // Read the current latest client state from the host chain store.
     let old_client_state = ctx.client_state(&client_id)?;
 
+    // Check if the client is frozen.
     if old_client_state.is_frozen() {
-        return Err(ClientError::ClientFrozen { client_id }.into());
+        return Err(ContextError::ClientError(ClientError::ClientFrozen {
+            client_id,
+        }));
```

Comment:
ibc-go doesn't check if the client is frozen (unless I missed it). There probably is a good reason for it, such as allowing a chain to "restart" clients on other chains? Not sure.

Reply:
Good point, Could be the reason! 
So, upgrade process would also allow restarting a frozen client, but I am not sure if it is a safe permission! 🤔

Reply:
#391